src/actions/hover.rs

```diff
@@ -0,0 +1,480 @@
+                        let the_type = format_method(ctx, contextstr.into());
+                        if !the_type.is_empty() {
+                            ty = Some(the_type.into())
+                        }
+                    },
+                    racer::MatchType::Trait | racer::MatchType::Enum | racer::MatchType::Struct => {
+                        let the_type = format_object(ctx, contextstr.into());
+                        if !the_type.is_empty() {
+                            ty = Some(the_type.into())
+                        }
+                    },
+                    _ => {
+                        if !contextstr.trim().is_empty() {
+                            ty = Some(contextstr.into())
+                        }
+                    }
+                }
+                trace!("hover_racer: racer_ty: {:?}", ty);
+                (m.docs, ty)
+            })
+            .next()
+    });
+
+    let results = results.map_err(|_| {
+        warn!("hover_racer: racer panicked");
+    });
+
+    results.unwrap_or(None)
+}
+
+/// Formats a struct, enum, union, or trait. The original type is returned
+/// in the event of an error.
+fn format_object(ctx: &InitActionContext, the_type: String) -> String {
+    trace!("hover: format_object: {}", the_type);
+    let config = ctx.fmt_config().get_rustfmt_config().clone();
+    let object = format!("{}{{}}", the_type);
+    let mut out = Vec::<u8>::with_capacity(the_type.len());
+    match format_input(FmtInput::Text(object), &config, Some(&mut out)) {
+        Ok(_) => {
+            let utf8 = String::from_utf8(out);
+            if let Ok((Some(pos), lines)) = utf8.map(|lines| (lines.rfind("{"), lines)) {
+                lines[0..pos].into()
+            } else {
+                the_type
+            }
+        },
+        Err(e) => {
+            warn!("hover: object format failed: {:?}", e);
+            the_type
+        }
+    }
+}
+
+/// Formats a method or function. The original type is returned
+/// in the event of an error.
+fn format_method(ctx: &InitActionContext, the_type: String) -> String {
+    trace!("hover: format_method: {}", the_type);
+    let config = ctx.fmt_config().get_rustfmt_config().clone();
+    let method = format!("impl Dummy {{ {} {{ unimplmented!() }} }}", the_type);
+    let mut out = Vec::<u8>::with_capacity(the_type.len());
+    match format_input(FmtInput::Text(method), &config, Some(&mut out)) {
+        Ok(_) => {
+            if let Ok(mut lines) = String::from_utf8(out) {
+                if let Some(front_pos) = lines.find("{") {
+                    lines = lines[1+front_pos..].into();
+                }
+                if let Some(back_pos) = lines.rfind("{") {
+                    lines = lines[0..back_pos].into();
+                }
+                lines.lines().filter(|line| line.trim() != "").map(|line| {
+                    format!("{}\n", &line[config.tab_spaces()..])
+                }).collect()
```

Comment:
Is the trailing '\n' wanted for vscode here, it causes unsightliness in atom

![before](https://user-images.githubusercontent.com/2331607/42566639-16faba7a-84ff-11e8-8b31-f739f039e649.png)

Removing the trailing newline fixes this
```rust
...
let mut s: String = lines
    .lines()
    .filter(|line| !line.trim().is_empty())
    .map(|line| format!("{}\n", &line[config.tab_spaces()..]))
    .collect();
s.pop();
s
```
![after](https://user-images.githubusercontent.com/2331607/42566653-1f815dd4-84ff-11e8-831f-e15bd8e8e526.png)